pkg/querier/lazy_querier.go

```diff
@@ -11,26 +11,36 @@ import (
 )
 
 type lazyQuerier struct {
+	ctx  context.Context
 	next storage.Querier
 }
 
 // newLazyQuerier wraps a storage.Querier, does the Select in the background.
 // Return value cannot be used from more than one goroutine simultaneously.
-func newLazyQuerier(next storage.Querier) storage.Querier {
-	return lazyQuerier{next}
+func newLazyQuerier(ctx context.Context, next storage.Querier) storage.Querier {
+	return lazyQuerier{ctx: ctx, next: next}
 }
 
 func (l lazyQuerier) Select(params *storage.SelectParams, matchers ...*labels.Matcher) (storage.SeriesSet, storage.Warnings, error) {
 	future := make(chan storage.SeriesSet)
 	go func() {
 		set, _, err := l.next.Select(params, matchers...)
+		var res storage.SeriesSet
 		if err != nil {
-			future <- errSeriesSet{err}
+			res = errSeriesSet{err}
 		} else {
-			future <- set
+			res = set
+		}
+		select {
+		case future <- res:
+			// ok
+		case <-l.ctx.Done():
+			// stop
 		}
```

Comment:
Could you explain why we need this and what it does? Isn't `l.next.Select()` blocking?

Reply:
You mean entire lazy set thing? There is comment about it in commit 76cc27152feeed512434a8a8506311f0e2099bbd, where it was added.

Now the problem with cancellation is that if one such querier returns error, another one will never get its `Next` method called, so this goroutine trying to send result would not be able to do anything. Now thinking about it, much simpler solution would be to make channel with buffer size of 1. I will change this PR to do that instead. Much simpler change, and fixes the same problem.